### PR TITLE
feat(grpc): Prometheus interceptors

### PR DESCRIPTION
**Pull request by Proto007**
## What does this PR address?

The current implementation creates an http server in a reserved port where the metrics can be accessed.

## Before submitting:

- [X] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [X] Does the code follow BentoML's code style, both `make format` and `make lint` script have passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [X] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [ ] Did your changes require updates to the documentation? Have you updated
  those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [ ] Did you write tests to cover your changes?

### CODE REVIEW

File: bentoml/_internal/server/grpc/interceptors/prometheus.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+            name=service_name + "_request_latency",
+            documentation="API GRPC response latency",
+            labelnames=["api_name", "service_version", "grpc_response_code"],
+        )
+        self.metrics_request_in_progress = metrics_client.Gauge(
+            name=service_name + "_request_in_progress",
+            documentation="Total number of GRPC requests in progress now",
+            labelnames=["api_name", "service_version"],
+        )
+        self._is_setup = True
```
Comment (aarnphm) [inline bentoml/_internal/server/grpc/interceptors/prometheus.py:78]: No need to do this. Just initialize these inside __init__

File: bentoml/_internal/server/grpc_app.py [hunk 2]
```diff
@@ -128,21 +132,32 @@ def interceptors(self) -> list[aio.ServerInterceptor]:
...
+            from .grpc.interceptors.prometheus import PrometheusServerInterceptor
+
+            prometheus_interceptor = PrometheusServerInterceptor(
+                bento_service=self.bento_service
+            )
+            interceptors.append(prometheus_interceptor)  # type: ignore
+
+        interceptors.extend(
+            [
+                GenericHeadersServerInterceptor(),
```
Comment (aarnphm) [inline bentoml/_internal/server/grpc_app.py:147]: This should be at L135

File: bentoml/_internal/server/grpc_app.py [hunk 3]
```diff
@@ -128,21 +132,32 @@ def interceptors(self) -> list[aio.ServerInterceptor]:
...
             from .grpc.interceptors import AccessLogServerInterceptor
 
             access_logger = logging.getLogger("bentoml.access")
             if access_logger.getEffectiveLevel() <= logging.INFO:
-                interceptors.append(AccessLogServerInterceptor)
+                interceptors.append(AccessLogServerInterceptor())
 
         # add users-defined interceptors.
-        interceptors.extend(self.bento_service.interceptors)
+        interceptors.extend(map(lambda x: x(), self.bento_service.interceptors))
```
Comment (aarnphm) [inline bentoml/_internal/server/grpc_app.py:161]: No need to initialize at every step. Just add the class and keep what it is.

File: bentoml/_internal/server/grpc_app.py (hunk 2 above)
Comment (Proto007) [reply bentoml/_internal/server/grpc_app.py:147]: Shouldn't prometheus interceptor be the first item on that list?

Comment (aarnphm) [reply bentoml/_internal/server/grpc_app.py:147]: GenericHeaders setups content-type and some others grpc headers

File: bentoml/_internal/server/grpc_app.py [hunk 4]
```diff
@@ -128,21 +132,32 @@ def interceptors(self) -> list[aio.ServerInterceptor]:
...
-            AsyncOtelInterceptor,
-        ]
+        interceptors: list[aio.ServerInterceptor] = []
+
+        if self.enable_metrics:
+            from .grpc.interceptors.prometheus import PrometheusServerInterceptor
+
+            prometheus_interceptor = PrometheusServerInterceptor(
+                bento_service=self.bento_service
+            )
```
Comment (aarnphm) [inline bentoml/_internal/server/grpc_app.py:142]: You can make a partial function for this.
```suggestion
            prometheus_interceptor = partial(PrometheusServerInterceptor, bento_service=self.bento_service)
```
import partial from functools.

File: bentoml/_internal/server/grpc/interceptors/prometheus.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    from ..types import Response
+    from ..types import RpcMethodHandler
+    from ..types import AsyncHandlerMethod
+    from ..types import HandlerCallDetails
+    from ..types import BentoServicerContext
+    from ....service import Service
+    from ...metrics.prometheus import PrometheusClient
+
+logger = logging.getLogger(__name__)
+START_TIME_VAR: "contextvars.ContextVar[float]" = contextvars.ContextVar(
```
Comment (aarnphm) [inline bentoml/_internal/server/grpc/interceptors/prometheus.py:28]:
```suggestion
START_TIME_VAR: contextvars.ContextVar[float] = contextvars.ContextVar(
```

File: bentoml/_internal/server/grpc/interceptors/prometheus.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        # ref: https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels
+        self.service_name = self.service_name.replace("-", ":").replace(".", "::")
+
+        self.service_version = (
+            bento_service.tag.version if bento_service.tag is not None else ""
+        )
+
+        self.metrics_request_duration = metrics_client.Histogram(
+            name=self.service_name + "_request_duration_seconds",
+            documentation=self.service_name + " API GRPC request duration in seconds",
```
Comment (aarnphm) [inline bentoml/_internal/server/grpc/interceptors/prometheus.py:57]: use f-string instead. Makes it easier to read. Applies for all of the below metrics.

File: bentoml/_internal/service/service.py [hunk 7]
```diff
@@ -246,9 +246,7 @@ def add_asgi_middleware(
     ) -> None:
         self.middlewares.append((middleware_cls, options))
 
-    def add_grpc_interceptor(
-        self, interceptor_cls: t.Type[grpc.aio.ServerInterceptor]
-    ) -> None:
+    def add_grpc_interceptor(self, interceptor_cls: grpc.aio.ServerInterceptor) -> None:
```
Comment (aarnphm) [inline bentoml/_internal/service/service.py:249]: Revert back this changes.

File: bentoml/_internal/server/instruments.py [hunk 8]
```diff
@@ -1,3 +1,5 @@
+from __future__ import annotations
```
Comment (aarnphm) [inline bentoml/_internal/server/instruments.py:1]: Maybe lets not include this in the `grpc` PR, since this has nothing to do with gRPC.

You can make a separate PR to `main` for types updates.

Comment (Proto007) [reply bentoml/_internal/server/instruments.py:1]: got it

File: bentoml/_internal/server/grpc/interceptors/prometheus.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+                # instrument request total count
+                self.metrics_request_total.labels(
+                    api_name=api_name,
+                    service_version=self.service_version,
+                    grpc_response_code=context.code(),
+                ).inc()
+
+                # instrument request duration
+                assert START_TIME_VAR.get() != 0
```
Comment (aarnphm) [inline bentoml/_internal/server/grpc/interceptors/prometheus.py:96]: What is the assert here are for?

File: bentoml/_internal/server/grpc/interceptors/prometheus.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+                total_time = max(default_timer() - START_TIME_VAR.get(), 0)
+                self.metrics_request_duration.labels(  # type: ignore
```
Comment (aarnphm) [inline bentoml/_internal/server/grpc/interceptors/prometheus.py:98]: What is the reason for `type: ignore` here?

File: bentoml/_internal/server/grpc/interceptors/prometheus.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+        self.metrics_request_duration = metrics_client.Histogram(
+            name=f"{self.service_name}_request_duration_seconds",
+            documentation=f"{self.service_name} API GRPC request duration in seconds",
+            labelnames=["api_name", "service_version", "grpc_response_code"],
+        )
+        self.metrics_request_total = metrics_client.Counter(
+            name=f"{self.service_name}_request_total",
+            documentation="Total number of GRPC requests",
+            labelnames=["api_name", "service_version", "grpc_response_code"],
```
Comment (aarnphm) [inline bentoml/_internal/server/grpc/interceptors/prometheus.py:61]: Instead of `grpc_response_code`, can we do `http_response_code` here? It is easier understand.

File: bentoml/_internal/server/grpc/interceptors/prometheus.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+                    grpc_response_code=context.code(),
+                ).inc()
+
+                # instrument request duration
+                assert START_TIME_VAR.get() != 0
+                total_time = max(default_timer() - START_TIME_VAR.get(), 0)
+                self.metrics_request_duration.labels(  # type: ignore
+                    api_name=api_name,
+                    service_version=self.service_version,
+                    grpc_response_code=context.code(),
```
Comment (aarnphm) [inline bentoml/_internal/server/grpc/interceptors/prometheus.py:101]: same with `grpc_response_code` here.

File: bentoml/_internal/server/grpc/interceptors/prometheus.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+                ).observe(total_time)
+                START_TIME_VAR.set(0)
```
Comment (aarnphm) [inline bentoml/_internal/server/grpc/interceptors/prometheus.py:103]: If this is the case then I don't think we should use contextvar here.

File: bentoml/_internal/server/grpc/interceptors/prometheus.py [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+                # instrument request total count
+                self.metrics_request_total.labels(
+                    api_name=api_name,
+                    service_version=self.service_version,
+                    grpc_response_code=context.code(),
+                ).inc()
+
+                # instrument request duration
+                assert START_TIME_VAR.get() != 0
+                total_time = max(default_timer() - START_TIME_VAR.get(), 0)
```
Comment (aarnphm) [inline bentoml/_internal/server/grpc/interceptors/prometheus.py:97]: Timer should wrap the rpc calls, so I believe this is not correct.
Use `try..finally` loop.

File: bentoml/_internal/server/grpc/interceptors/prometheus.py (hunk 9 above)
Comment (Proto007) [reply bentoml/_internal/server/grpc/interceptors/prometheus.py:96]: I was following this from `instruments.py`. I think it's to make sure the request started. The following line:
https://github.com/bentoml/BentoML/blob/c0b500988c3575b87fc1e65d5b4d9b3a7c535c17/bentoml/_internal/server/instruments.py#L102

File: bentoml/_internal/server/grpc/interceptors/prometheus.py (hunk 10 above)
Comment (Proto007) [reply bentoml/_internal/server/grpc/interceptors/prometheus.py:98]: Without the `type:ignore`, pylance complains:
`Type of "observe" is partially unknown`
`Type of "observe" is "(amount: Any, exemplar: Any | None = None) -> None"`

Comment (aarnphm) [reply bentoml/_internal/server/grpc/interceptors/prometheus.py:99]: then you should add something like `# type: ignore (unfinished prometheus types)`

adding some stubs afterwards makes it easier to know the reason behind the ignore.

**Review by aarnphm — APPROVED** [review]
